### DISCUSSION TITLE
Faster SQL Bulk Inserts With C#

### DISCUSSION THREAD

**Discussion by NETSphereSoft** [timdeschryver/timdeschryver.dev#254] [Announcements]
On your page https://github.com/timdeschryver/timdeschryver.dev/blob/main/blog/faster-sql-bulk-inserts-with-csharp/index.md I think there is a comparison missing, handling inserts in a table with a `DbCommand` and `DbParameter`.  
  
Instead creating a DbCommand and it's parameters each time there should be use a single `DbCommand` instance and it's list of already created parameters. The `ExecuteNonQuery()` can be executed unlimited.

```cs

var customerItem = customers.First();

using (var connection = new SqlConnection(connectionString))
using (var command = new SqlCommand(cmdText, connection);)
{
    command.Parameters.AddWithValue("@Id", customerItem.Id);
    command.Parameters.AddWithValue("@FirstName", customerItem.FirstName);
    command.Parameters.AddWithValue("@LastName", customerItem.LastName);
    command.Parameters.AddWithValue("@Street", customerItem.Street);
    command.Parameters.AddWithValue("@City", customerItem.City);
    command.Parameters.AddWithValue("@State", customerItem.State);
    command.Parameters.AddWithValue("@PhoneNumber", customerItem.PhoneNumber);
    command.Parameters.AddWithValue("@EmailAddress", customerItem.EmailAddress);

    connection.Open();

    foreach (var customer in customers)
    {
        command.Parameters[0].Value = customer.Id;
        command.Parameters[1].Value = customer.FirstName;
        command.Parameters[2].Value = customer.LastName;
        command.Parameters[3].Value = customer.Street;
        command.Parameters[4].Value = customer.City;
        command.Parameters[5].Value = customer.State;
        command.Parameters[6].Value = customer.PhoneNumber;
        command.Parameters[7].Value = customer.EmailAddress;
 
        command.ExecuteNonQuery();
    }
}
```

Bit other kind of that:

```cs
var first = true;

using (var connection = new SqlConnection(connectionString))
using (var command = new SqlCommand(cmdText, connection);)
{
    var parameters = command.Parameters;

    connection.Open();

    foreach (var customer in customers)
    {
        if(first)
        {
            parameters.AddWithValue("@Id", customer.Id);
            parameters.AddWithValue("@FirstName", customer.FirstName);
            parameters.AddWithValue("@LastName", customer.LastName);
            parameters.AddWithValue("@Street", customer.Street);
            parameters.AddWithValue("@City", customer.City);
            parameters.AddWithValue("@State", customer.State);
            parameters.AddWithValue("@PhoneNumber", customer.PhoneNumber);
            parameters.AddWithValue("@EmailAddress", customer.EmailAddress);

            first = false;
        }
        else
        {
            parameters[0].Value = customer.Id;
            parameters[1].Value = customer.FirstName;
            parameters[2].Value = customer.LastName;
            parameters[3].Value = customer.Street;
            parameters[4].Value = customer.City;
            parameters[5].Value = customer.State;
            parameters[6].Value = customer.PhoneNumber;
            parameters[7].Value = customer.EmailAddress;
        }

        command.ExecuteNonQuery();
    }
}
```

**Comment by NETSphereSoft** [comment]
I think there is a comparison missing, handling inserts in a table with a `DbCommand` and `DbParameter`.  
  
Instead creating a DbCommand and it's parameters each time there should be use a single `DbCommand` instance and it's list of already created parameters. The `ExecuteNonQuery()` can be executed unlimited.

```cs

var customerItem = customers.First();

using (var connection = new SqlConnection(connectionString))
using (var command = new SqlCommand(cmdText, connection);)
{
    command.Parameters.AddWithValue("@Id", customerItem.Id);
    command.Parameters.AddWithValue("@FirstName", customerItem.FirstName);
    command.Parameters.AddWithValue("@LastName", customerItem.LastName);
    command.Parameters.AddWithValue("@Street", customerItem.Street);
    command.Parameters.AddWithValue("@City", customerItem.City);
    command.Parameters.AddWithValue("@State", customerItem.State);
    command.Parameters.AddWithValue("@PhoneNumber", customerItem.PhoneNumber);
    command.Parameters.AddWithValue("@EmailAddress", customerItem.EmailAddress);

    connection.Open();

    foreach (var customer in customers)
    {
        command.Parameters[0].Value = customer.Id;
        command.Parameters[1].Value = customer.FirstName;
        command.Parameters[2].Value = customer.LastName;
        command.Parameters[3].Value = customer.Street;
        command.Parameters[4].Value = customer.City;
        command.Parameters[5].Value = customer.State;
        command.Parameters[6].Value = customer.PhoneNumber;
        command.Parameters[7].Value = customer.EmailAddress;
 
        command.ExecuteNonQuery();
    }
}
```

Bit other kind of that:

```cs
var first = true;

using (var connection = new SqlConnection(connectionString))
using (var command = new SqlCommand(cmdText, connection);)
{
    var parameters = command.Parameters;

    connection.Open();

    foreach (var customer in customers)
    {
        if(first)
        {
            parameters.AddWithValue("@Id", customer.Id);
            parameters.AddWithValue("@FirstName", customer.FirstName);
            parameters.AddWithValue("@LastName", customer.LastName);
            parameters.AddWithValue("@Street", customer.Street);
            parameters.AddWithValue("@City", customer.City);
            parameters.AddWithValue("@State", customer.State);
            parameters.AddWithValue("@PhoneNumber", customer.PhoneNumber);
            parameters.AddWithValue("@EmailAddress", customer.EmailAddress);

            first = false;
        }
        else
        {
            parameters[0].Value = customer.Id;
            parameters[1].Value = customer.FirstName;
            parameters[2].Value = customer.LastName;
            parameters[3].Value = customer.Street;
            parameters[4].Value = customer.City;
            parameters[5].Value = customer.State;
            parameters[6].Value = customer.PhoneNumber;
            parameters[7].Value = customer.EmailAddress;
        }

        command.ExecuteNonQuery();
    }
}
```

The approach above I have more maximized, additionally you can use this in combination with multi threading - `Parallel`, this makes it easy to handle thousands of records in seconds. My best tries I had with 24 threads in parallel, but not with your data. 150'000 small records were done in ~4 seconds. In `MySQL` with an approach of `DbDataAdapter` from `MySqlConnector` I could insert in ~1.2 seconds.

**Reply by timdeschryver** [reply]
I wasn't aware that this could be done. Thanks for sharing this knowledge @NETSphereSoft !
(I'm also going to update your comment to include the snippets so it's also available on the blog page).

**Comment by akuma6099** [comment]
Just wanted to say thanks for this writeup. I'm scraping 100's of millions of records from an environment and inserting them 1-by-1 was a horrible solution. Now I'm at around 20-30 seconds per half-million because the dataset is wider than your samples. Thanks a bunch!

**Reply by timdeschryver** [reply]
I'm happy that this was useful, thanks for the feedback @akuma6099 !

**Comment by henrSea** [comment]
Love this. However, looking at the WriteToServer call "ToDataTable" and not seeing this as available option. could be I'm running an older version of .Net? Or perhaps I overlooked creating this method in your code?  Cheers

**Reply by timdeschryver** [reply]
@henrSea It seems like I forgot to mention that this is a method we created, to map a list to a datatable. Sorry about that 😅

**Reply by MikeGriffinReborn** [reply]
```csharp
static DataTable ToDataTable(List<Customer> customers)
{
    DataTable dataTable = new DataTable();

    var col_Id = dataTable.Columns.Add("Id",  typeof(int));
    var col_FirstName = dataTable.Columns.Add("FirstName", typeof(String));
    var col_LastName = dataTable.Columns.Add("LastName", typeof(String));
    var col_Street = dataTable.Columns.Add("Street", typeof(String));
    var col_City = dataTable.Columns.Add("City", typeof(String));
    var col_State= dataTable.Columns.Add("State", typeof(String));

    foreach (var customer in customers)
    {
        DataRow row = dataTable.NewRow();

        dataTable.Rows.Add(row);
        row[col_FirstName] = customer.FirstName;
        row[col_LastName] = customer.LastName;
        row[col_Street] = customer.Street;
        row[col_City] = customer.City;
        row[col_State] = customer.State;
    }

    return dataTable;
}
```

**Comment by VenkataAnand114** [comment]
Hi @Tim, Could you share the code for ToDataTable() Function that is being used?

**Reply by MikeGriffinReborn** [reply]
```csharp
static DataTable ToDataTable(List<Customer> customers)
{
    DataTable dataTable = new DataTable();

    var col_Id = dataTable.Columns.Add("Id",  typeof(int));
    var col_FirstName = dataTable.Columns.Add("FirstName", typeof(String));
    var col_LastName = dataTable.Columns.Add("LastName", typeof(String));
    var col_Street = dataTable.Columns.Add("Street", typeof(String));
    var col_City = dataTable.Columns.Add("City", typeof(String));
    var col_State= dataTable.Columns.Add("State", typeof(String));

    foreach (var customer in customers)
    {
        DataRow row = dataTable.NewRow();

        dataTable.Rows.Add(row);
        row[col_FirstName] = customer.FirstName;
        row[col_LastName] = customer.LastName;
        row[col_Street] = customer.Street;
        row[col_City] = customer.City;
        row[col_State] = customer.State;
    }

    return dataTable;
}
```

**Comment by MaulikDarji-aaa** [comment]
Hey Tim, 
This is really great comparison. 
I am working on the Table parameter option as I have a SP which checks if need to insert or update the row in SQL. 
I am facing a challenge with that. If the column sequence in the tableType does not match with dataTable, it stores value in the columns sent by dataTable. 
Is there a solution for this?

**Reply by timdeschryver** [reply]
Sorry @MaulikDarji-aaa , I think this is the behavior of using table types. 
As far as I know there is no workaround for this.

**Comment by psyciknz** [comment]
Hi.  Great write up.  

For Generated-sql downwards.   how do you support an issue with the insertion of one record when inserting in builk.

I'm contemplating a move away from SQLBulkCopy due to the error reporting it has.  When there's an issue with one field in a record, and that record is part of a large dataset.  You might get some column information (generally hard to resolve back to the input database), but then you get no information as to which record.  Unless you use the SqlRowsCopied callback with a fairly low resolution to find where it got up to.

**Reply by timdeschryver** [reply]
That might be one of the reasons not to use SQLBulkCopy.
Thanks for adding this information.

**Reply by psyciknz** [reply]
Isn't the same problem likely with all of the methods, short of looping insert statements?

**Reply by Notheisz57** [reply]
> Isn't the same problem likely with all of the methods, short of looping insert statements?

If incompatibilities between your data stores is a common occurrence, then your scenario is leaving the territory of what can be found in .NET.
Solutions exist to provide flexible/resilient data integrations.  (ex. SQL Server Integration Services)

That said, if the errors are very uncommon, a reasonable solution might be to use a `SqlTransaction`:
 - [SqlBulkCopy(SqlConnection, SqlBulkCopyOptions, SqlTransaction)](https://learn.microsoft.com/en-us/dotnet/api/microsoft.data.sqlclient.sqlbulkcopy.-ctor?view=sqlclient-dotnet-core-5.2#microsoft-data-sqlclient-sqlbulkcopy-ctor(microsoft-data-sqlclient-sqlconnection-microsoft-data-sqlclient-sqlbulkcopyoptions-microsoft-data-sqlclient-sqltransaction))

If the transaction lets you roll-back a partial/failed insertion, the failing record could be isolated using a dynamic/variable batch size.
I have not tried this myself, but I would consider an approach like this:
- Insert records in batches, say 100000 records initially.  (Keep the current batch size in a variable)
- Whenever an insertion fails, try again with 1/10th the batch size. (10000, 1000, ...)
- But, if an insertion fails and your batch size was only 1, this is the failing record.  Store it somewhere, then advance to the next batch.
- Whenever an insertion succeeds, double the current batch size, unless the current batch size is equal to initial/desired batch size.

**Comment by Martino-Bordin** [comment]
Hi Tim,
thank you for this article.
I just want to say that actually DbSet.AddRange DO perform Change detection, but just one time.

Quoting the doc:

By default, Entity Framework performs Detect Changes automatically when the following methods are called:

DbSet.Find
DbSet.Local
DbSet.Add
**DbSet.AddRange**
DbSet.Remove
DbSet.RemoveRange
DbSet.Attach
DbContext.SaveChanges
DbContext.GetValidationErrors
DbContext.Entry
DbChangeTracker.Entries

--

**If you need to add or remove many objects from a context, consider using DbSet.AddRange and DbSet.RemoveRange. This methods automatically detect changes only once after the add or remove operations are completed.**
